specs/service_deduction.adoc

```diff
@@ -0,0 +1,174 @@
+// vim: tw=79
+
+= Service Deduction Module for Node Agent
+
+Node agents need to ship with modules that can detect various services already
+present on the nodes
+
+== Problem description
+
+This specification describes service deduction module to manage the services
+present on the nodes and populate the details; to and from the central store.
+The list could then be presented to the administrator via the API and the UI
+to allow selection of the nodes for monitoring, service operation and
+troubleshoting.
+
+== Use Cases
+
+* Tendrl should detects the service details of the storage node and show
+the details and pushes the details to central store for further use.
+
+* Use Case1: Monitoring
+  ** The list of available nodes presented to the administrator via APIs
+  ** User selects one node in the cluster
+  ** From the selected node user selects the deducted services
+  ** From the selected service user can watch the status
+     or submit request to start or stop or reload the service
+
+* Use Case2: Trouble-shooting
+  ** The list of available nodes presented to the administrator via APIs
+  ** User selects one node in the cluster
+  ** From the selected node user selects the deducted services
+  ** From the selected service user submit request to start or stop or
+     reload the service
+
+* Use Case3: During import
+  ** The list of available nodes presented to the administrator via APIs
+  ** User selects one node in the cluster
+  ** Use the selected node to extract the details of the cluster which
+     uses deducted service status
+
+== Proposed change
+
+Node Agents related changes:
+* Add a module which fetches the service details like service-name, execution
+path, executing script, status, how much cpu percentage it takes, how much
+memory it consumes, number of io counts on read and write, open files,
+number of threads.
+
+* Create a new class for service details in persistence module and
+  add required functions and update all necessary changes in the node-agent
+  manager class.
+
+* Deduct the services details from the tendrl-service-module and update
+  on any changes into centerl store.
+
+* Create an object in tendrl_definitions_node_agent for service details.
+
+=== Alternatives
+None
+
+=== Data model impact:
+* Add service details in centrl store with keys
+  ** nodes/node_id/Services/status/service_name
+  ** nodes/node_id/Services/stop/service_name
+  ** nodes/node_id/Services/start/service_name
+  ** nodes/node_id/Services/restart/service_name
+  ** nodes/node_id/Services/reload/service_name
+* Add service object in tendrl_definitions_node_agent/service
+
+=== Impacted Modules:
+
+==== Tendrl API impact:
+
+Service status flow requires the below APIs to be exposed:
+
+* API to Get the service details
+
+----
+
+GET /1.0/GetServices
```

Comment:
I don't think you need a top level APi for this? Can we embed these details as part of GetNodeList itself?